### PR TITLE
`EmptyStatement`

### PR DESCRIPTION
**Pull request by imteekay**
Implement the empty statement token
- [ecmascript: Empty Statement](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-empty-statement)

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -23,12 +23,12 @@ export type Lexer = {
...
-  Assignment,
-  ExpressionStatement,
-  Var,
-  TypeAlias,
+  Identifier = 'Identifier',
+  Literal = 'Literal',
+  Assignment = 'Assignment',
+  ExpressionStatement = 'ExpressionStatement',
+  Var = 'Var',
+  TypeAlias = 'TypeAlias',
```
Comment (sandersn) [inline src/types.ts:31]: You need a new kind of node here, `EmptyStatement`

File: src/parse.ts [hunk 2]
```diff
@@ -69,6 +75,8 @@ export function parse(lexer: Lexer): Module {
       parseExpected(Token.Equals);
       const typename = parseIdentifier();
       return { kind: Node.TypeAlias, name, typename, pos };
+    } else if (tryParseToken(Token.EOF)) {
+      return undefined;
```
Comment (sandersn) [inline src/parse.ts:79]: and you should return `{ kind: Node.EmptyStatement }` here.

File: src/parse.ts [hunk 3]
```diff
@@ -69,6 +75,8 @@ export function parse(lexer: Lexer): Module {
       parseExpected(Token.Equals);
       const typename = parseIdentifier();
       return { kind: Node.TypeAlias, name, typename, pos };
+    } else if (tryParseToken(Token.EOF)) {
```
Comment (sandersn) [inline src/parse.ts:78]: EOF won't work if you have an EmptyStatement in the middle of other statements: `var x: number = 1;;var y: number = 2`. I think you'll have to peek at the lexer state with `lexer.token()` to see if it's a semicolon OR EOF.

File: src/test.ts [hunk 4]
```diff
@@ -100,7 +100,7 @@ function display(o: any) {
   const o2 = {} as any;
   for (const k in o) {
     if (k === 'pos') continue;
-    else if (k === 'kind') o2[k] = Node[o.kind];
+    else if (k === 'kind') o2[k] = Node[o.kind as Node];
     else if (typeof o[k] === 'object') o2[k] = display(o[k]);
     else o2[k] = o[k];
   }
```
Comment (sandersn) [inline src/test.ts:106]: be sure to add a test and its baselines to show that EmptyStatements now appear.

File: src/parse.ts [hunk 5]
```diff
@@ -69,10 +69,20 @@ export function parse(lexer: Lexer): Module {
       parseExpected(Token.Equals);
       const typename = parseIdentifier();
       return { kind: Node.TypeAlias, name, typename, pos };
+    } else if (tryParseTokens([Token.EOF, Token.Semicolon])) {
```
Comment (sandersn) [inline src/parse.ts:72]: the tricky part is that you don't actually want to *parse* the semicolon. That consumes it, but the parseSeparated code is supposed to do that. You just need to look ahead and see that the next token is semicolon/EOF. That's enough to know that you have EmptyStatement.

This means that empty statements are zero width. They are literally empty.

Comment (imteekay) [reply src/parse.ts]: Thanks for the feedback, now I got the idea. https://github.com/imteekay/mini-typescript/pull/2/commits/6a4834cca50234f7ebe92b8581d0794cf9b6b5d4

Comment (imteekay) [reply src/parse.ts]: Thanks for the feedback @sandersn, now I got the idea. https://github.com/imteekay/mini-typescript/pull/2/commits/6a4834cca50234f7ebe92b8581d0794cf9b6b5d4

**Comment by imteekay** [comment]
@sandersn, do you think this solution/implementation fits the `EmptyStatement` exercise?

**Review by sandersn — APPROVED** [review]
Looks good! A couple more suggestions as well.

File: src/emit.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { Statement, Node, Expression } from './types';
 
 export function emit(statements: Statement[]) {
-  return statements.map(emitStatement).join(';\n');
+  return statements.filter(Boolean).map(emitStatement).join(';\n');
```
[the remark this answers is not in the document]
Comment (imteekay) [reply src/emit.ts]: That makes sense, just removed the filter https://github.com/imteekay/mini-typescript/pull/2/commits/6182029a7b46b74fc4a2301baaeeff1d33b67a7e

File: src/check.ts [hunk 7]
```diff
@@ -13,6 +13,7 @@ import { resolve } from './bind';
 const stringType: Type = { id: 'string' };
 const numberType: Type = { id: 'number' };
 const errorType: Type = { id: 'error' };
+const emptyType: Type = { id: 'empty' };
```
[the remark this answers is not in the document]
Comment (imteekay) [reply src/check.ts]: Looking at the [spec](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-empty-statement-runtime-semantics-evaluation), it says it returns `empty`. Just renamed from `emptyType` to `empty` https://github.com/imteekay/mini-typescript/pull/2/commits/6da8ef07f47646d816c6a259250a0f18b681bdba Lmk if this is what you meant for the refactoring.

**Pull request merged**